server/testCases/src/index.ts

```diff
@@ -0,0 +1,163 @@
+import { createServer } from "http";
+import { CHALLENGE_LOCALHOST_PORT, LOCALHOST_PORT } from "./constants";
+import { VM } from "vm2";
+import { TestCase, Challenge, TestCaseResults } from "./types";
+
+dotenv.config();
+
+const app = express();
+const server = createServer(app);
+app.use(express.json());
+
+const MAX_TIMEOUT = 2000; // 2 seconds
+
+/**
+ * Fetches and validates a challenge by ID
+ * @param challengeId The ID of the challenge to fetch
+ * @returns The challenge data or error
+ */
+const fetchChallenge = async (challengeId: string) => {
+  try {
+    const response = await fetch(
+      `http://localhost:${CHALLENGE_LOCALHOST_PORT}/challenge/${challengeId}`
+    );
+
+    if (!response.ok) {
+      return response.status === 404
+        ? { error: "Challenge id not found" }
+        : { error: `Failed to fetch challenge: ${response.statusText}` };
+    }
+
+    const data = await response.json();
+
+    if (!Array.isArray(data.testCases) || data.testCases.length === 0) {
+      return { error: "No test cases found for this challenge" };
+    }
+
+    return { challenge: data };
+  } catch (error) {
+    return { error: `Failed to fetch challenge: ${(error as Error).message}` };
+  }
+};
+
+/**
+ * Tests a function against a set of test cases
+ * @param func The function to be tested
+ * @param testCases the test cases to be run
+ * @returns error or results
+ */
+const runTestCase = (func: string, testCase: TestCase) => {
+  try {
+    const { input, output, id, difficultyWeight } = testCase;
+
+    const inputString = JSON.stringify(input);
+
+    let userOutput: any;
+    let error = null;
+
+    try {
+      const vm = new VM({
+        timeout: MAX_TIMEOUT, // Automatically terminates if too slow
+        sandbox: {},
+      });
+      const result = vm.run(`
+          const func = (...args) => {${func}};
+          func(...${inputString});
+        `);
+
+      userOutput = result;
+    } catch (error) {
+      error = `Error in test execution: ${(error as Error).message}`;
+    }
+
+    if (error) return { error };
+
+    const passed = JSON.stringify(userOutput) === JSON.stringify(output);
+
+    const testResult = {
+      id,
+      input,
+      expectedOutput: output,
+      actualOutput: userOutput,
+      passed,
+      difficultyWeight,
+    };
+
+    return { result: testResult };
+  } catch (error) {
+    return { error: `Error running test cases: ${(error as Error).message}` };
+  }
+};
+
+const runTestCases = async (func: string, challengeId: Challenge["id"]) => {
+  const { challenge, error } = await fetchChallenge(challengeId);
+  if (error) return { error };
+
+  const { testCases } = challenge;
+  const total = challenge.testCases.length;
+
+  let totalDifficultyWeight = 0;
+  let testCasesPassed = 0;
+
+  const testCaseResults: TestCaseResults["results"] = [];
+
+  try {
+    for (const testCase of testCases) {
+      const { result, error } = runTestCase(func, testCase);
```

Comment:
wrapping this in a try catch is bit complicated and unnecessary imo. You designed the runTestCase to catch all errors and return them in the error field by default so this whole try catch is redundant. Instead type the return of runTestCase to be an object that could either return one field or the other and use typescripts built in narrowing to the need for non-null assertions and unnest the logic